bundle/edu.gemini.sp.vcs/src/main/scala/edu/gemini/sp/vcs/diff/ObsNumberCorrection.scala

```diff
@@ -0,0 +1,84 @@
+package edu.gemini.sp.vcs.diff
+
+import edu.gemini.pot.sp.SPNodeKey
+import edu.gemini.sp.vcs.diff.NodeDetail.Obs
+import edu.gemini.sp.vcs.diff.ProgramLocation.{Remote, Local}
+import edu.gemini.spModel.rich.pot.sp._
+
+import scalaz._
+import Scalaz._
+
+/** Corrects observation numbers such that all local-only observations have
+  * numbers greater than the maximum remote observation number.
+  *
+  * There are a few implicit assumptions at work here.  First, once an
+  * observation has been created its key is always remembered in the
+  * `VersionMap` (that's true of any node actually).  Second, as new
+  * observations are created they are given the next sequential integer.
+  * That is, observation numbers aren't reused when deleted.  Finally
+  * observation numbers are never modified once assigned (other than by
+  * the merge plan application itself).
+  *
+  * This means that there cannot exist an observation known to both local and
+  * remote program versions with an observation number greater than a
+  * local-only observation.
+  */
+class ObsNumberCorrection(isKnown: (ProgramLocation, SPNodeKey) => Boolean) extends MergeCorrection {
+  def apply(mp: MergePlan): MergePlan = {
+    val t = (mp.update.loc/:renumberedObs(mp)) { case (loc, (key, num)) =>
+      loc.find(_.tree.rootLabel.key === key).fold(loc) { obsLoc =>
+        val mergeNode = obsLoc.tree.rootLabel match {
+          case Modified(k,nv,dob,Obs(_)) => Modified(k, nv, dob, Obs(num))
+          case lab                       => lab
+        }
+        obsLoc.setLabel(mergeNode).root
+      }
+    }.toTree
```

Comment:
I'm not pleased with this bit.  Here we take the `Tree[MergeNode]` and for each observation to be renumbered, we find it in the tree and update the observation number.  Instead of repeatedly finding nodes in the tree from the root, it seems like we should be able to walk the tree applying corrections but every attempt I made was awkward.